rust/yarp-sys/tests/parser_tests.rs

```diff
@@ -0,0 +1,212 @@
+        let comment_list = &parser.comment_list;
+        let comment = comment_list.head as *const yp_comment_t;
+        assert_eq!((*comment).type_, yp_comment_type_t::YP_COMMENT_INLINE);
+
+        let location = {
+            let start = (*comment).start.offset_from(parser.start);
+            let end = (*comment).end.offset_from(parser.start);
+            start..end
+        };
+        assert_eq!(location, 0..7);
+
+        yp_node_destroy(parser, _node);
+        yp_parser_free(parser);
+    }
+}
+
+#[test]
+fn diagnostics_test() {
+    let source = CString::new("class Foo;").unwrap();
+    let mut parser = MaybeUninit::<yp_parser_t>::uninit();
+
+    unsafe {
+        yp_parser_init(
+            parser.as_mut_ptr(),
+            source.as_ptr(),
+            source.as_bytes().len(),
+            std::ptr::null(),
+        );
+        let parser = parser.assume_init_mut();
+        let _node = yp_parse(parser);
+
+        let error_list = &parser.error_list;
+        let error = error_list.head as *const yp_diagnostic_t;
+        let message = CStr::from_ptr((*error).message);
+        assert_eq!(
+            message.to_string_lossy(),
+            "Expected to be able to parse an expression."
+        );
+
+        let location = {
+            let start = (*error).start.offset_from(parser.start);
+            let end = (*error).end.offset_from(parser.start);
+            start..end
+        };
+        assert_eq!(location, 10..10);
+
+        yp_node_destroy(parser, _node);
+        yp_parser_free(parser);
+    }
+}
+
+#[test]
+fn encoding_change_test() {
+    unsafe extern "C" fn callback(_parser: *mut yp_parser_t) {
+        eprintln!("hi");
+        THING.get_mut().map(|v| {
+            *v += 1;
+            v
+        });
+    }
+
+    static mut THING: OnceCell<u8> = OnceCell::new();
+    unsafe { THING.set(42).unwrap() };
+
+    let source = CString::new(
+        "# encoding: ascii\nclass Foo; end
+    ",
+    )
+    .unwrap();
+    let mut parser = MaybeUninit::<yp_parser_t>::uninit();
+
+    unsafe {
+        yp_parser_init(
+            parser.as_mut_ptr(),
+            source.as_ptr(),
+            source.as_bytes().len(),
+            std::ptr::null(),
+        );
+        let parser = parser.assume_init_mut();
+
+        yp_parser_register_encoding_changed_callback(parser, Some(callback));
+
+        let _node = yp_parse(parser);
+        assert!(parser.encoding_changed);
+
+        // This value should have been mutated inside the callback when the encoding changed.
+        assert_eq!(*THING.get().unwrap(), 43);
+
+        yp_node_destroy(parser, _node);
+        yp_parser_free(parser);
+    }
+}
+
+#[test]
+fn encoding_decode_test() {
+    unsafe extern "C" fn callback(
+        _parser: *mut yp_parser_t,
+        name: *const c_char,
+        width: usize,
+    ) -> *mut yp_encoding_t {
+        let c_name = CStr::from_ptr(name);
+        assert_eq!(c_name.to_string_lossy(), "meow");
+        assert_eq!(width, 4);
```

Comment:
I suppose I've never encountered asserting inside an `extern "C" fn`... The docs there seem a bit confusing to me. On initial read, I'm inclined to change this, but then https://doc.rust-lang.org/beta/nomicon/ffi.html#panic-can-be-stopped-at-an-abi-boundary makes it sound like my assert calls would be fine. And, in fact, if I change one of those assertions, I get a regular ol' Rusty failure:

```
failures:

---- encoding_decode_test stdout ----
thread 'encoding_decode_test' panicked at 'assertion failed: `(left == right)`
  left: `"meow"`,
 right: `"pants"`', tests/parser_tests.rs:181:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

...which (admittedly, naïvely) seems like it's ok?

Reply:
this behavior is currently being changed as part of the rollout of the `C-unwind` ABI. In future Rust, this will turn into an `abort(3)` which will forcibly terminate the test process.

Reply:
Solved by using another icky `OnceLock`.